pvlib/horizon.py

```diff
@@ -0,0 +1,646 @@
+
+    interpolated_elevs = interpolator(pts).reshape(total_num_samples, 1)
+    samples = np.hstack((pts, interpolated_elevs))
+    return samples
+
+
+def uniformly_sample_triangle(p1, p2, p3, num_samples):
+    """
+    Randomly sample the surface of a triangle defined by three (lat, lon, elev)
+    points uniformly [1].
+
+    Parameters
+    ----------
+    pt1 : ndarray
+        An array conaining (lat, lon, elev) values that define one vertex
+        of the triangle.
+    pt2 : ndarray
+        An array conaining (lat, lon, elev) values that define another vertex
+        of the triangle.
+    pt3 : ndarray
+        An array conaining (lat, lon, elev) values that define the last vertex
+        of the triangle.
+
+    num_samples : tuple
+        The number of random samples to be uniformly taken from the surface
+        of the triangle.
+
+    Returns
+    -------
+    points: Nx3 ndarray
+        Array with N (lat, lon, elev) points that lie on the surface of the
+        triangle.
+
+    [1] http://graphics.stanford.edu/courses/cs468-08-fall/pdf/osada.pdf
+    """
+    c1 = tools.lle_to_xyz(p1)
+    c2 = tools.lle_to_xyz(p2)
+    c3 = tools.lle_to_xyz(p3)
+
+    r1 = np.random.rand(num_samples).reshape((num_samples, 1))
+    r2 = np.random.rand(num_samples).reshape((num_samples, 1))
+    sqrt_r1 = np.sqrt(r1)
+
+    random_pts = (1-sqrt_r1)*c1 + sqrt_r1*(1-r2)*c2 + sqrt_r1*r2*c3
+    random_pts = tools.xyz_to_lle(random_pts)
+
+    return random_pts
+
+
+def filter_points(horizon_azimuths, horizon_angles, bin_size=1):
+    """
+    Bins the horizon_points by azimuth values. The azimuth value of each
+    point in horizon_points is rounded to the nearest bin and then the
+    max value in each bin is returned.
+
+    Parameters
+    ----------
+    horizon_azimuths: numeric
+        Azimuth values for points that define the horizon profile. The ith
+        element in this array corresponds to the ith element in horizon_angles.
+
+    horizon_angles: numeric
+        Elevation angle values for points that define the horizon profile. The
+        elevation angle of the horizon is the angle that the horizon makes with
+        the horizontal. It is given in degrees above the horizontal. The ith
+        element in this array corresponds to the ith element in
+        horizon_azimuths.
+
+    bin_size : int
+        The width of the bins for the azimuth values.
+
+    Returns
+    -------
+    filtered_azimuths: numeric
+        Azimuth values for points that define the horizon profile. The ith
+        element in this array corresponds to the ith element in
+        filtered_angles.
+
+    filtered_angles: numeric
+        elevation angle values for points that define the horizon profile given
+        in degrees above the horizontal. The ith element in this array
+        corresponds to the ith element in filtered_azimuths.
+
+    """
+    assert(horizon_azimuths.shape[0] == horizon_angles.shape[0])
+
+    wedges = {}
```

Comment:
I'll guess this loop may be a bit slow. What about instead:
```
rounded_azimuths = tools.round_to_nearest(horizon_azimuths, bin_size)
bins = np.unique(rounded_azimuths)

filtered = np.column_stack((bins, np.nan * bins))

for i in range(filtered.shape[0]):
    idx = np.nonzero(rounded_azimuths == filtered[i, 0])
    filtered[i, 1] = np.max(horizon_angles[idx])

return filtered[:, 0], filtered[:, 1]
```

Now that I've thought that through, there will always be a bin at 0 degrees, so the user ought to be informed of this point. I'd add to the note about `bin_size`.

Reply:
A couple of questions on this.
Is `np.nonzero` necessary?  I think you can just index with the boolean array. 
Why will there always be a bin at 0 degrees?

Reply:
> A couple of questions on this.
> Is np.nonzero necessary?  I think you can just index with the boolean array.

You're probably right. I started with [argwhere](https://docs.scipy.org/doc/numpy/reference/generated/numpy.argwhere.html#numpy.argwhere), decided it wasn't the right approach, stopped at `nonzero` without thinking further.

> Why will there always be a bin at 0 degrees?

You are right again, there won't always be a bin at 0.  But the bins will be a subset of 0, +/-bin_size, +/-2*bin_size, etc.  That's what I meant to describe and what I think we shoudl tell the user.

Reply:
Needs tests for the functions added to `tools.py` then I'm OK with the review.